src/VirtoCommerce.Platform.Caching/Redis/RedisPlatformMemoryCache.cs

```diff
@@ -1,47 +1,98 @@
         private readonly ILogger _log;
-        private readonly RetryPolicy _retryPolicy;
+        private readonly TelemetryClient _telemetryClient;
 
         private static readonly string _cacheId = Guid.NewGuid().ToString("N");
         
 
         public RedisPlatformMemoryCache(IMemoryCache memoryCache, IOptions<CachingOptions> options
+            , ConnectionMultiplexer connection
             , ISubscriber bus
+            , IOptions<CachingOptions> cachingOptions
             , IOptions<RedisCachingOptions> redisCachingOptions
             , ILogger<RedisPlatformMemoryCache> log
+            , TelemetryClient telemetryClient
             ) : base(memoryCache, options, log)
        {
+            _memoryCache = memoryCache;
             _log = log;
+            _telemetryClient = telemetryClient;
             _bus = bus;
 
+            _cachingOptions = cachingOptions.Value;
             _redisCachingOptions = redisCachingOptions.Value;
-            _bus.Unsubscribe(_redisCachingOptions.ChannelName);
-            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage);
+
+            connection.ConnectionFailed += OnConnectionFailed;
+            connection.ConnectionRestored += OnConnectionRestored;
+
+            _bus.Subscribe(_redisCachingOptions.ChannelName, OnMessage, CommandFlags.FireAndForget);
 
             _log.LogInformation($"{nameof(RedisPlatformMemoryCache)}: subscribe to channel {_redisCachingOptions.ChannelName } current instance:{ _cacheId }");
+            _telemetryClient.TrackEvent("RedisSubscribed", new Dictionary<string, string>
+            {
+                {"channelName", _redisCachingOptions.ChannelName},
+                {"cacheId", _cacheId}
+            });
+        }
+
+        protected virtual void OnConnectionFailed(object sender, ConnectionFailedEventArgs e)
+        {
+            _log.LogError($"Redis disconnected from instance { _cacheId }. Endpoint is {e.EndPoint}, failure type is {e.FailureType}");
+            _telemetryClient.TrackException(e.Exception);
+            _telemetryClient.TrackEvent("RedisDisconnected", new Dictionary<string, string>
+            {
+                {"channelName", _redisCachingOptions.ChannelName},
+                {"cacheId", _cacheId},
+                {"endpoint", e.EndPoint.ToString()},
+                {"failureType", e.FailureType.ToString()}
+            });
 
-            _retryPolicy = Policy.Handle<Exception>().WaitAndRetry(
-                _redisCachingOptions.BusRetryCount, retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt - 1)));
+            // If we have no connection to Redis, we can't invalidate cache on another platform instances,
+            // so the better idea is to disable cache at all for data consistence
+            CacheEnabled = false;
+            // We should fully clear cache because we don't know
+            // what's changed until platform found Redis is unavailable
+            GlobalCacheRegion.ExpireRegion();
```

Comment:
The code like this cannot be tested or this hard to do. Potentially dangerous.
Can we use these cache manipulations with singleton service which is defined by DI container?

Reply:
It can be tested easy, because regions are cancellable.
Why it's potentially dangerous? Did you read PR description?

Reply:
Yeah, I've read.
Potentially dangerous cuz we using a self-defined singleton with static fields and methods. Better to use DI for this.
Eg: _globalCacheService.ExpireRegion();

Maybe it is out of scope, but anyway

Reply:
How are you going to test singleton with static fields? Do you want to write a special wrapper for it?

Reply:
All cache regions are static. It's by design.